### DISCUSSION TITLE
How to solve Exception: No gradient attached to this NDArray

### DISCUSSION THREAD

**Discussion by kameronren** [deepjavalibrary/djl#1041] [Q&A] [answered]
**djl version: 0.11.0**
first i write a function: `sgd(net.getParameters(), (float)0.2, batchSize);` but when I call it, jump out a Exception(**No gradient attached to this NDArray, please call array.requiredGradient()on your NDArray or block.setInitializer() on your Block**);

so, second I writed a other function: 
`public static void requiresGrad(ParameterList params) {
	    for (int i = 0; i < params.size(); i++) {	
	    	NDArray param = params.get(i).getValue().getArray();
	    	param.setRequiresGradient(true);
	    }
}`. but when I call it( **requiresGrad(net.getParameters());** ), jump out a other **Exception(The array has not been initialized)**.

so,third I change the function(**requiresGrad**) :

`public static void requiresGrad(ParameterList params) {
	    for (int i = 0; i < params.size(); i++) {		
	    	params.get(i).getValue().setInitializer(new NormalInitializer());
	    	params.get(i).getValue().initialize(manager, DataType.FLOAT32);
	    	NDArray param = params.get(i).getValue().getArray();
	    	System.out.println(params.get(i).getValue().isInitialized());
	    	param.setRequiresGradient(true);
	    }
}`. but when I call it ( **requiresGrad(net.getParameters());** ),jump out a third **Exception( java.lang.NullPointerException: No parameter shape has been set)**

how to solve the question? thanks!

**Comment by frankfliu** [comment]
Please take a look D2L book: https://d2l.djl.ai/chapter_preliminaries/autograd.html

You can get latest version here: https://github.com/deepjavalibrary/d2l-java/blob/master/chapter_preliminaries/autograd.ipynb

**Comment by kameronren** [comment, marked as answer]
I kown why **requiresGrad** function is not work!  The key point is  to put zhe function after the `NDList results = net.forward(parameterStore, X, true);` If put the **requiresGrad** function before `NDList results = net.forward(parameterStore, X, true);` .then we get the **Exception(The array has not been initialized)**
